Fix receipt broadcast to listening clients; add unit tests

jimthematrix opened the pull request:
Fixing the broadcast of transaction receipts to listening clients.

A client can listen for receipts on websocket by connecting to the `ws://<host>:<port>/ws` endpoint. It will be registered with the fabconnect server once it sends the following payload:

```
{"type":"listenreplies"}
```

peterbroadhurst commented on internal/rest/receipt/receiptstore.go:
```diff
@@ -92,11 +83,17 @@ func NewReceiptStore(config *conf.RESTGatewayConf) ReceiptStore {
+		r.persistence = persistence
+		return nil
+	} else {
+		return r.persistence.Init()
```
Trying to understand this - maybe a comment would be useful?

It seems like the caller of this asserts that `r.persistence` is **not** `nil`, by passing `nil` in for the `persistence` parameter.

jimthematrix replied:
added inline comments for clarifying these branches (only for testing purposes)

peterbroadhurst commented on internal/rest/receipt/receiptstore.go:
```diff
@@ -174,7 +171,7 @@ func (r *receiptStore) writeReceipt(requestID string, receipt map[string]interfa
 		existing, qErr := r.persistence.GetReceipt(requestID)
 		if qErr == nil && existing != nil {
-			log.Warnf("%s: exiting   receipt: %+v", requestID, *existing)
+			log.Warnf("%s: existing   receipt: %+v", requestID, *existing)
```
Note a whitespace mismatch after the change

jimthematrix replied:
fixed

peterbroadhurst commented on internal/rest/receipt/receiptstore.go:
```diff
@@ -201,12 +198,6 @@ func (r *receiptStore) GetReceipts(res http.ResponseWriter, req *http.Request, p
 		return
 	}
 
-	res.Header().Set("Content-Type", "application/json")
```
Why is this no longer required?

Is this explicitly done in `marshalAndReply`?

jimthematrix replied:
indeed, moved to `marshalAndReply` for both this and `GetReceipt`

peterbroadhurst reviewed (approved):
couple of minor comments to review, along with approval

jimthematrix commented:
I was getting the following error when trying to push a commit for addressing the review comments. So will leave them to a later PR.

```
remote: error: GH006: Protected branch update failed for refs/heads/receipt-broadcast.
remote: error: Required status check "DCO" is expected.
To github.com:hyperledger/firefly-fabconnect.git
 ! [remote rejected] receipt-broadcast -> receipt-broadcast (protected branch hook declined)
error: failed to push some refs to 'github.com:hyperledger/firefly-fabconnect.git'
```
